src/base/tag.h

```diff
@@ -0,0 +1,70 @@
+/*
+ * Bittorrent Client using Qt and libtorrent.
+ * Copyright (C) 2023  Vladimir Golovnev <glassez@yandex.ru>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#pragma once
+
+#include <compare>
+
+#include <QMetaType>
+#include <QString>
+
+class Tag final
+{
+public:
+    Tag() = default;
+
+    explicit Tag(const QString &tagStr);
+    explicit Tag(const std::string &tagStr);
+
+    bool isValid() const;
+    bool isEmpty() const;
+
+    QString toString() const noexcept;
+
+    bool operator==(const Tag &) const = default;
+    explicit operator QString() const noexcept;
+
+private:
+    QString m_tagStr;
+};
+
+Q_DECLARE_METATYPE(Tag)
+
+QDataStream &operator<<(QDataStream &out, const Tag &tag);
+QDataStream &operator>>(QDataStream &in, Tag &tag);
+
+std::size_t qHash(const Tag &key, std::size_t seed = 0);
```

Comment:
> You don't need this since you already provided `std::hash` specialization.

Unlike `std::hash` `qHash` has a more rich/advanced interface (second parameter `seed`), so I thought it would be nice to have its explicit overload for use in the context of Qt, and leave `std::hash` only for compatibility with the standard library.

Reply:
> You should provide the `seed` to std::hash specialization.
> https://doc.qt.io/qt-6/qhash.html#the-hashing-function

Hmm... I didn't find such a signature on [C++ reference](https://en.cppreference.com/w/cpp/utility/hash/operator()).
On the other hand, if this parameter is optional, then it simply does not affect the code that does not know about it.

Reply:
Seems you (accidentally?) reverted some other changes along the way:
https://github.com/qbittorrent/qBittorrent/compare/e5589b505a807cb93f9d3400169217b03a5ea171..de778562640ec7049958a9cf00e6a229e2fc86be

Reply:
> Seems you (accidentally?) reverted some other changes along the way

Thank you! Fixed.